### PR TITLE
Implement Node.js compat process.env

### PR DESCRIPTION
**Pull request by jasnell**
Exposes text bindings from the worker via process.env

Requires internal changes and CI also.

### CODE REVIEW

File: src/workerd/io/worker.c++ [hunk 1]
```diff
@@ -1350,7 +1348,18 @@ Worker::Worker(kj::Own<const Script> scriptParam,
...
+      if (script->getIsolate().getApiIsolate().getFeatureFlags().getNodeJsCompat()) {
+        processEnv = v8::Object::New(lock.v8Isolate);
+      }
+
+      compileBindings(lock, *script->isolate->apiIsolate, bindingsScope, processEnv);
+
+      KJ_IF_MAYBE(env, processEnv) {
+        auto key = v8::Private::ForApi(lock.v8Isolate,
+                                       jsg::v8StrIntern(lock.v8Isolate, "processEnv"));
+        KJ_REQUIRE(jsg::check(context->Global()->SetPrivate(context, key, *env)));
```
Comment (jasnell) [inline src/workerd/io/worker.c++:1361]: Note: done this way, the `process.env` object is mutable at the global scope, which matches Node.js' behavior. However, mutating `process.env` does *not* mutate the existing `env` argument passed into the handler, or used at the global scope, preserving the existing behavior. When documenting `process.env`, we'll just need to notate that `process.env` is independently mutable.

File: src/workerd/server/workerd-api.c++ [hunk 2]
```diff
@@ -629,26 +629,28 @@ static v8::Local<v8::Value> createBindingValue(
...
   auto context = lock.v8Context();
   auto& featureFlags = *impl->features;
 
   for (auto& global: globals) {
-    v8::HandleScope scope(lock.v8Isolate);
-
     // Don't use String's usual TypeHandler here because we want to intern the string.
-    auto name = jsg::v8StrIntern(lock.v8Isolate, global.name);
+    auto name = jsg::v8StrIntern(isolate, global.name);
     auto value = createBindingValue(lock, global, featureFlags);
```
Comment (jasnell) [inline src/workerd/server/workerd-api.c++:643]: Complete side note: I have to wonder if there's any value at all in lazily instantiating bindings in terms of reducing cold start times. Probably not significant enough to spend the time implementing but something we could probably investigate later?

File: src/workerd/server/workerd-api.c++ [hunk 3]
```diff
@@ -629,26 +629,28 @@ static v8::Local<v8::Value> createBindingValue(
...
     uint32_t ownerId) const {
   auto& lock = kj::downcast<JsgWorkerdIsolate::Lock>(lockParam);
-  v8::HandleScope scope(lock.v8Isolate);
+  auto isolate = lock.v8Isolate;
+  v8::HandleScope scope(isolate);
   auto context = lock.v8Context();
   auto& featureFlags = *impl->features;
 
   for (auto& global: globals) {
-    v8::HandleScope scope(lock.v8Isolate);
```
Comment (jasnell) [inline src/workerd/server/workerd-api.c++:639]: A couple drive-by improvements made while I was in here. There's no reason to enter a new v8::HandleScope on each iteration since we already entered one at the top of the loop.

**Comment by kentonv** [comment]
Unfortunately, I don't think this works. In Modules syntax, multiple instances of `Worker` with different environments may share the same isolate and global scope. With this implementation, each time a new `Worker` is created, it'll clobber `process.env` causing the previous `Worker` (which may still be running) to see the wrong environment.

**Comment by jasnell** [comment]
Yeah I forgot about that detail. I've updated the PR to account for that.

Specifically there are two distinct `process.env` scopes: global-scope and per-worker instance. At the global scope, the `process.env` will initially return an empty object but the worker script can mutate it. At the per-worker instance scope, the `process.env` will inherit properties from the global scope and will include vars from the binding configuration.

So, for instance, let's say I have the worker script:

```
import { env } from 'node:process';

env.FOO = '123';
env.BAR = 'XYZ';

export default {
  async fetch(request) {
    console.log('env.FOO', env.FOO);
    console.log('env.BAR', env.BAR);
    return new Response("ok");
  }
};
```

And let's suppose I have two different workers configured to use the same script, one with a `FOO` binding whose value is 'AAA', and another whose `FOO` binding is `BBB`, within the fetch call for each instance, the values of `env.FOO` and `env.BAR` will be `AAA`/`XYZ`, and `BBB`/`XYZ` respectively.

File: src/workerd/api/node/node.h [hunk 4]
```diff
@@ -29,6 +30,53 @@ class CompatibilityFlags : public jsg::Object {
   }
 };
 
+class Process : public jsg::Object {
+public:
+  v8::Local<v8::Object> getEnv(jsg::Lock& js, bool rootContext = true) {
```
Comment (kentonv) [inline src/workerd/api/node/node.h:35]: Nitpick, but there's a lot of code in these inline functions, which will end up being compiled as part of workerd-api.c++ (and the corresponding internal file) because that's the place that actually calls it, making that file take even longer to compile. Moving the code out-of-line helps avoid that.

**Comment by kentonv** [comment]
Seems fine as discussed (or at least, I can't think of a better idea), will leave detailed review to the tagged reviewers.

**Comment by jasnell** [comment]
One question that comes up is what the order of precedence should be when setting a value explicitly at global scope. Specifically, if user code explicitly sets `process.env.FOO = 'abc'` at the global scope and there is a `FOO` text binding set to `'XYZ'`, then in the request the configured binding will currently take precedence. Should it?

File: src/workerd/api/node/node.h (hunk 4 above)
Comment (jasnell) [reply src/workerd/api/node/node.h:35]: Fair point :-)

**Comment by jasnell** [comment]
Ok, based on discussion and evaluation I've updated this to remove the automatic per-request population of the `process.env` object. It is going to be a bit tricky to get that done cleanly so punting on that for now.

**Comment by irvinebroque** [comment]
This is only exposed when within a `NodeJsModule` type — correct?

**Comment by jasnell** [comment]
`process` is a global in the `NodeJsModule` type, `node:process` can be imported anywhere when the nodeJsCompat flag is enabled.

**Review by ohodson — APPROVED** [review]
